File: truevfs-kernel/src/main/java/de/schlichtherle/truevfs/kernel/se/NeedsLockRetryException.java
/*
 * Copyright (C) 2005-2012 Schlichtherle IT Services.
 * All rights reserved. Use is subject to license terms.
 */
package de.schlichtherle.truevfs.kernel.se;

import javax.annotation.Nullable;
import javax.annotation.concurrent.Immutable;

/**
 * Indicates that all file system locks need to get released before the
 * operation can get retried.
 *
 * @see    FsSyncController
 * @author Christian Schlichtherle
 */
@Immutable
@SuppressWarnings("serial") // serializing a control flow exception is nonsense!
final class NeedsLockRetryException extends ControlFlowException {

    private static final @Nullable NeedsLockRetryException
            SINGLETON = TRACEABLE ? null : new NeedsLockRetryException();

    static NeedsLockRetryException get() {
        return TRACEABLE ? new NeedsLockRetryException() : SINGLETON;
    }

    private NeedsLockRetryException() { }
}

File: truevfs-kernel/src/main/scala/de/schlichtherle/truevfs/kernel/se/NeedsLockRetryException.scala
/*
 * To change this template, choose Tools | Templates
 * and open the template in the editor.
 */

package de.schlichtherle.truevfs.kernel.se

/**
 * Indicates that all file system locks need to get released before the
 * operation can get retried.
 *
 * @see    SyncController
 * @author Christian Schlichtherle
 */
private final class NeedsLockRetryException private()
extends ControlFlowException

private object NeedsLockRetryException {
  import ControlFlowException._

  private[this] val instance = new NeedsLockRetryException

  def apply() = if (traceable) new NeedsLockRetryException else instance
}
